// message_service/client.js
var http = require('http');

var queue = require(__dirname + '/queue');
var msgmod = require(__dirname + '/message');

// main way to create a client
var createClient = function(config) {
    return new Client(config);
}

//
var Client = function(config) {
    this.config = config;

    var self = this;

    // send a message
    this.sendMessage = function(message) {
        // build a request object
        var reqConf = {
            host: this.config.host,
            port: this.config.port,
            method: 'POST'
        };

        var request = http.request(reqConf, function(response) {
            console.log('STATUS: ' + res.statusCode);
            console.log('HEADERS: ' + JSON.stringify(res.headers));
            res.setEncoding('utf8');
            res.on('data', function (chunk) {
                console.log('BODY: ' + chunk);
            });
        });

        request.on('error', function(e) {
            console.log('problem with request: ' + e.message);
            console.log('retrying message in 3 seconds: ' + message.toString());
            setTimeout(self.sendMessage(message), 3000);
        });

        //write the message
        request.write(message.payload);
        //signal end of request
        request.end();
    }
}


module.exports = {
    createClient: createClient
};
