// message_service/service.js
var net = require('net');
var server = require('server');
var events = require('events');
//var client = require('client');

// config structure:
// see http://nodejs.org/api/net.html#net_net_createconnection_options_connectionlistener
// {
//     host: 'foo.example.com',
//     port: 1234,
//     localAddress: '127.0.0.1'
// }


var createService = function(config) {
    return new Service(config);
}


var Service = function(config) {
    if(typeof config === 'undefined') config = 
        {
            host: '127.0.0.1'
            ,localAddress: 0
            ,port: 8124
        };
    this.config = config;
};


Service.prototype.start = function() {
    this.server = server.createServer(this.config);
    this.server.start();
    this.clients = [];
}


Service.prototype.shutdown = function() {
    this.server.shutdown();
};

Service.prototype.addFriend = function(config) {
    this.clients.push(createClient(config));
};

Service.prototype.sendMessage = function(msg) {
    console.log('sending message');
    clients = this.clients;
    for(var x in clients) {
        // todo: wrap message and use a queue
        console.log('client '+x);
        clients[x].write(msg);
    }
};


var createClient = function(config) {
    var client =  net.createConnection(config, function() {
        console.log('client connected');
    });

    client.on('data', function(data) {
        console.log(data.toString());
        client.end();
    });

    client.on('end', function() {
        console.log('client disconnected');
    });

    client.on('error', function(err) {
        console.error(err);
    });

    return client;
}


module.exports = {
    createService: createService,
};
