// message_service/service.js
var net = require('net');
var events = require('events');
var util = require('util');

var server = require(__dirname + '/server');
var message = require(__dirname + '/message');
var client = require(__dirname + '/client');

var EVENT_RECEIVED_MESSAGE = 'received_message';

// config structure:
// see http://nodejs.org/api/net.html#net_net_createconnection_options_connectionlistener
// {
//     host: 'foo.example.com',
//     port: 1234,
//     localAddress: '127.0.0.1'
// }


var createService = function(config) {
    return new Service(config);
}


var Service = function(config) {
    if(typeof config === 'undefined') config = 
        {
            host: '127.0.0.1'
            ,localAddress: 0
            ,port: 8124
        };
    this.config = config;

    events.EventEmitter.call(this);
};
util.inherits(Service, events.EventEmitter);


Service.prototype.start = function() {
    var self = this;
    this.server = server.createServer(this.config);

    // callback to propagate message arrival
    this.server.on(server.EVENT_RECEIVED_MESSAGE, function(message) {
        console.log('service receieved data:'+message);
        self.emit(EVENT_RECEIVED_MESSAGE, message);
    });

    this.server.start();
    this.clients = [];
}


Service.prototype.shutdown = function() {
    this.server.shutdown();
};

Service.prototype.addFriend = function(config) {
    this.clients.push(client.createClient(config));
};

Service.prototype.sendMessage = function(data) {
    console.log('sending message');
    var msg = new message.Message(data);

    clients = this.clients;
    for(var x in clients) {
        // todo: wrap message and use a queue
        console.log('client '+x);
        clients[x].write(msg.toMessage());
    }
};


module.exports = {
    createService: createService,
    EVENT_RECEIVED_MESSAGE: EVENT_RECEIVED_MESSAGE
};
