// message_service/server.js
var net = require('net');
var events = require('events');
var util = require('util');

var parser = require(__dirname + '/message_parser');

var EVENT_RECEIVED_MESSAGE = 'received_message';

// config structure:
// see http://nodejs.org/api/net.html#net_net_createconnection_options_connectionlistener
// {
//     port: 1234,
//     localAddress: '127.0.0.1'
// }

var createServer = function(config) {
    return new Server(config);
}


var Server = function(config) {
    if(typeof config === 'undefined') config = {localAddress: 0, port: 8124};
    this.config = config;
    
    events.EventEmitter.call(this);
};
util.inherits(Server, events.EventEmitter);


Server.prototype.start = function() {
    var self = this;
    console.log('starting server');
    this.server = net.createServer(function(c) { //'connection' listener
        console.log('server connected');

        c.on('end', function() {
            console.log('server disconnected');
        });

        c.on('readable', function() {
            messages = parser.parse(c.read());
            for(x in messages) {
                console.log('server receieved data:'+messages[x]);
                self.emit(EVENT_RECEIVED_MESSAGE, messages[x]);
            }
        });

        c.write('hello\r\n');
        c.pipe(c);
    });

    this.server.listen(this.config.port, this.config.localAddress, function() { //'listening' listener
      console.log('server bound');
    });
};

// this will stop new connections from being accepted
// and fire the callback when shutdown is complete
Server.prototype.shutdown = function() {
    console.log('shutting down server');
    this.server.close(function() {
        console.log('shutdown complete');
    });
};

Server.prototype.getServer = function() {
    return this.server;
};


module.exports = {
    createServer: createServer,
    EVENT_RECEIVED_MESSAGE: EVENT_RECEIVED_MESSAGE
};
